### PR TITLE
Update javascript loop guidelines

### PR DESCRIPTION
**Pull request by basil-conto**
##### Reviewers (feel free to add more)
- [x] @Xevi 
- [x] @marzetas 
- [x] @io41 
- [x] @alvaroortega 
- [x] @full-of-foo

### CODE REVIEW

File: frontend.md [hunk 1]
```diff
@@ -102,16 +102,66 @@ Object.keys(person).forEach(function(prop){
...
+```javascript
+// Old style for loop
+var found = false;
+for (var i = 0; !found && i < array.length; i++) {
+    found = (array[i] === 5);
+}
+
+// New declarative style
+var found = array.some(function(x) {
+    return (x === 5);
```
Comment (full-of-foo) [inline frontend.md]: Nitpick: I would omit the brackets here, so `return x === 5`

File: frontend.md [hunk 2]
```diff
@@ -102,16 +102,66 @@ Object.keys(person).forEach(function(prop){
...
+In cases where looping functions from `Array.prototype` like the aforementioned
+are impractical and a plain loop is needed, prefer `for` loops over `while`
+loops if the number of iterations is known, e.g. when iterating over an array:
+
+```javascript
+// Short-circuited while loop
+var i = 0;
+var found = false;
+while (!found && i < array.length) {
+    found = (array[i] === 5);
```
Comment (full-of-foo) [inline frontend.md]: Ditto

File: frontend.md [hunk 3]
```diff
@@ -102,16 +102,66 @@ Object.keys(person).forEach(function(prop){
...
+    i++;
+}
+
+console.log(found ? 'Found at index ' + (i - 1) : 'Not found');
+
+// For loops are more concise and capture the control logic in a single place,
+// making it harder to, say, forget to update the counter
+var found = false;
+for (var i = 0; !found && i < array.length; i++) {
+    found = (array[i] === 5);
```
Comment (full-of-foo) [inline frontend.md]: Ditto

File: frontend.md [hunk 4]
```diff
@@ -102,16 +102,66 @@ Object.keys(person).forEach(function(prop){
...
+
+// Contrary to languages with similar syntax, the counter variable has the same
+// scope as the loop, so this will work:
+console.log(found ? 'Found at index ' + (i - 1) : 'Not found');
+
+// Although this can, in fact, be written declaratively, the gain in clarity is
+// arguably little to none:
+var index = 0;
+found = array.some(function(x, i) {
+    return (x === 5) && ((index = i) === i);
```
Comment (full-of-foo) [inline frontend.md]: Ditto

File: frontend.md [hunk 5]
```diff
@@ -102,16 +102,66 @@ Object.keys(person).forEach(function(prop){
...
+});
+
+console.log(found ? 'Found at index ' + index : 'Not found');
```
Comment (marzetas) [inline frontend.md]: I wouldn't even mention a counter-example, just list the correct way to do things.

**Comment by full-of-foo** [comment]
+1 from me after mentioned comments

File: frontend.md (hunk 5 above)
Comment (full-of-foo) [inline frontend.md]: I agree

**Comment by alvaroortega** [comment]
+1

**Comment by Xevi** [comment]
I'm gonna miss the `while`

**Pull request merged**